### PR TITLE
Fix invariant-breaking parsing of errant case clauses in switch expressions

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/71957

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 1]
```diff
@@ -570,62 +570,62 @@ private SwitchExpressionSyntax ParseSwitchExpression(ExpressionSyntax governingE
...
+                parseSwitchExpressionArms(),
                 this.EatToken(SyntaxKind.CloseBraceToken));
-        }
-
-        private SeparatedSyntaxList<SwitchExpressionArmSyntax> ParseSwitchExpressionArms()
-        {
-            var arms = _pool.AllocateSeparated<SwitchExpressionArmSyntax>();
 
-            while (this.CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+            SeparatedSyntaxList<SwitchExpressionArmSyntax> parseSwitchExpressionArms()
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:576]: view with whitespace off.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler this is ready for review.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 2]
```diff
@@ -570,62 +570,62 @@ private SwitchExpressionSyntax ParseSwitchExpression(ExpressionSyntax governingE
...
+                        // Help out in the case where a user is converting a switch statement to a switch expression.
+                        // Consume the `:` as a `=>` and report an error.
+                        this.CurrentToken.Kind == SyntaxKind.ColonToken
+                            ? this.EatTokenAsKind(SyntaxKind.EqualsGreaterThanToken)
+                            : this.EatToken(SyntaxKind.EqualsGreaterThanToken),
+                        ParseExpressionCore());
+
+                    // If we're not making progress, abort
+                    if (errantCase is null && switchExpressionCase.FullWidth == 0 && this.CurrentToken.Kind != SyntaxKind.CommaToken)
+                        break;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:612]: the bug was here.  it was possible to consume the errant-case but then discard it, leading to it being dropped from the final tree.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @333fred ptal.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:1]: Can we get some versions with leading and trailing trivia like comments as well?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:1]: Sure!

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:1]: done.

**Comment by RikkiGibson** [comment]
It looks like tests are timing out, perhaps one of the new test cases is causing an infinite loop?

**Comment by CyrusNajmabadi** [comment]
@333fred good now?

**Pull request merged**